Optimizer package for Graph mode

Craigacp opened the pull request:
This adds a new subproject `tensorflow-training` which currently contains `org.tensorflow.training.optimizers` a package of gradient optimizers which apply the underlying gradient update ops to a TF Graph.

In addition to the optimizers it makes a few small changes in the `tensorflow-core-api` package: it adds a variables initialiser list to `Graph`, a method which constructs an initialiser node which initialises all the variables in the graph, plus a `variableWithInit` method which accepts an `Operand` which is used to provide the shape, type and initial value for the variable.

There's also an MNIST CNN test in there, which we can move out to wherever it needs to go, and be combined with the other ones.

Before this gets merged it needs better Javadoc in the Optimizer class, and the global variables used by some of the optimizers need wiring into the globals field in the base class. I'll fix those things next week.

karllessard commented:
@Craigacp , just quickly like that, it looks like you'll need to rebase your PR to resolve a conflict

dhruvrajan commented:
@Craigacp cool stuff! Did you mean to commit the files under `src/annotations/gen`? (those get regenerated at build time, if I'm not mistaken?)

Craigacp commented:
I rebased the branch and the deterministic generation commits went away with the associated conflict, but I seem to have picked up the ci changes that got merged in earlier today.

karllessard commented on tensorflow-training/src/main/java/org/tensorflow/training/optimizers/AdaDelta.java:
```diff
@@ -0,0 +1,92 @@
+
+  private final float rho;
+
+  private final float epsilon;
```
I don't think we need to enforce it but just FYI, pretty much all TF classes declare their members in this order, including both fields and methods: `public`, `protected`, `default`, `private`

Craigacp replied:
Do you mean all the public methods & fields, then protected etc, or all the public methods, protected methods, ... then public fields, protected fields etc?

karllessard commented on tensorflow-training/src/main/java/org/tensorflow/training/optimizers/Optimizer.java:
```diff
@@ -0,0 +1,245 @@
+
+  protected Optimizer(Graph graph) {
+    this.graph = graph;
+    this.tf = Ops.create(graph).withName(getOptimizerName());
```
would it be easy to allow the user override the name of the scope for the optimizer ops (instead of `getOptimizerName()`)? Also, maybe a user would like to pass its own instance of `Ops` (which might already be a subscope of another block or contains control dependencies, etc.)?

Craigacp replied:
It's locked off to `Graph` at the moment as I don't think any of this works in eager mode, so the type system enforces that. I can allow a hook for the name, and we could relax the `Graph` check to an `instanceof` check throwing `IllegalArgumentException` to supply an ops. Should I keep the Graph entry point as well?

karllessard replied:
Yes I think it is ok that to have an entry point that accept a `Graph` but maybe add another that accept an `Ops` as well, so the user has full control on the name and control dependencies of the optmizers? Or maybe just a `String` for the name

For instance, does it make sense that a user would want to create two different optimizers of the same type (but with different parameters) for handling different variables? If so, then he'll need to give them different names or he will endup with a name conflict.

I think you understand my questioning here, I'll let you decide what would be the best approach to handle those corner cases.

Craigacp replied:
I added an additional constructor which accepts a Graph and a String for the base name of the operations.

Craigacp commented on .github/workflows/ci.yml:
```diff
@@ -1,5 +1,11 @@
 name: CI jobs
-on: [push, pull_request]
+on:
```
It looks like Github has started squishing PR commits into a single one? Starting with the deterministic ops one and then the github actions one. Maybe that's got something to do with it.

I'll do another rebase along with the rest of the changes.

Craigacp commented on tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/op/core/VariableWithInit.java:
```diff
@@ -0,0 +1,67 @@
+
+// This class has been generated, DO NOT EDIT!
+
+package org.tensorflow.op.core;
```
Ok, I'll sort that. I think it would be better as an endpoint anyway.

Craigacp commented on tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/op/core/VariableWithInit.java:
```diff
@@ -0,0 +1,67 @@
+    if (exEnv instanceof Graph) {
+      Graph graph = (Graph) exEnv;
+      graph.addInitializer(assignOp);
+    }
```
I'll check.

Craigacp commented on tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java:
```diff
@@ -0,0 +1,331 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.tensorflow.training.examples;
```
I think it might be nice to have both the Logistic Regression MNIST you have, along with this more complex CNN example. Though we could just put them both in the same file, all the data loading, evaluation and training code will be the same.

Craigacp commented on tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java:
```diff
@@ -0,0 +1,331 @@
+    Relu<TFloat32> relu1 = tf.nn.relu(tf.nn.biasAdd(conv1, conv1Biases));
+
+    // First pooling layer
+    MaxPool<TFloat32> pool1 = tf.nn.maxPool(relu1, tf.constant(new int[]{1, 2, 2, 1}), tf.constant(new int[]{1, 2, 2, 1}), PADDING_TYPE);
```
Yep, I'm happy to rebase after the other two PRs have merged. I've not had chance to review the latest versions of either of them yet though.

Craigacp replied:
Done

Craigacp commented on tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java:
```diff
@@ -0,0 +1,331 @@
+    logger.info("Optimizer = " + optimizer.toString());
+    Op minimize = optimizer.minimize(loss, TRAIN);
+
+    Op init = graph.variablesInitializer();
```
Without calling `graph.variablesInitializer()` it doesn't actually create the init node on the graph, as it doesn't know when you've finished adding nodes. Similar to the minimize call above, these both mutate the graph to add appropriate entry points, but I don't need the ops they produce here (because they get special names).

Craigacp commented on tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java:
```diff
@@ -0,0 +1,331 @@
+  public static final String INIT = "init";
+
+  public static Graph build(String optimizerName) {
+    Graph graph = new Graph();
```
At L268? Sure. I'll do a pass through the test code and tidy these things up.

Craigacp commented on tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java:
```diff
@@ -0,0 +1,331 @@
+    float[][][][] trainData = train.features;
+    int[] trainLabels = train.labels;
+
+    float[][][][] testData = test.features;
```
That's because I load in multidimensional arrays from the serialised files I have lying around with MNIST in. When we move this over to the java-models repo I can use your MNIST loader which will get rid of this nonsense.

Craigacp replied:
Moved the file over to examples.

Craigacp commented on tensorflow-training/src/main/java/org/tensorflow/training/optimizers/AdaDelta.java:
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
```
We don't have an authors file in this repo, plus this is the text that my lawyers require on all outbound source contributions. I'm not sure if putting it in the AUTHORS file will be sufficient for them. It looks like the main TF repo has an AUTHORS, but then it says to look at CONTRIBUTORS which doesn't exist, so maybe someone at Google should figure out what they want it to look like for all the TF related repos.

Craigacp commented:
I've rebased on the latest master, fixed the style issues, put the try with resources in, and updated it to use the `@Endpoint` annotation for `variableWithInit`. I've also stubbed out using `@Endpoint` for the Adam optimizer, and I think it should work, but it would require the optimizers to be in the `tensorflow-core-api` project as otherwise there will be ugly circular dependencies. TODOs are update the optimizer constructor to allow an ops, rearrange the code so it's ordered like the other bits of the project, and check the behaviour of `tf.assign` in eager mode.

karllessard commented on tensorflow-training/src/main/java/org/tensorflow/training/optimizers/Optimizer.java:
```diff
@@ -0,0 +1,279 @@
+    List<Operation> variables = new ArrayList<>();
+    Iterator<Operation> opItr = graph.operations();
+    while (opItr.hasNext()) {
+      Operation op = opItr.next();
```
`graph.operations().forEachRemaining` maybe?

Also, is it OK that an optimizer is always applied to all variables in the graph? Is this true for all kind of graphs?

Craigacp replied:
That's how they work in Python. It's really hard to make it work any other way without a lot of ceremony.

karllessard commented on tensorflow-training/src/main/java/org/tensorflow/training/optimizers/Optimizer.java:
```diff
@@ -0,0 +1,279 @@
+      Variable<? extends TType> slot = variables.get(varName);
+      if (slot != null) {
+        @SuppressWarnings("unchecked") // This method should only be called when the type is known.
+            Optional<Variable<T>> opt = Optional.of((Variable<T>) slot);
```
Nit: either it is GitHub or there is a wrong left margin to this line. Also, no need of else blocks below since you return in previous blocks.

Craigacp replied:
The Google code formatter in IntelliJ keeps doing this, I'll keep moving it back. It's because it treats the annotation and the variable creation as the same statement, and thinks it's a line continuation.

karllessard commented on tensorflow-training/src/main/java/org/tensorflow/training/optimizers/RMSProp.java:
```diff
@@ -0,0 +1,117 @@
+          tf.constant(momentum, gradient.dataType()),
+          tf.constant(epsilon, gradient.dataType()),
+          gradient);
+    } else {
```
Nit: no need of else block

karllessard commented:
Please don’t forget to rebase your PR before pushing a new version as I just merged some other breaking changes (mainly renaming ˋtf.constant` to ˋtf.val` and adding ˋtf.array`), thanks

Craigacp commented on tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/CoreOps.java:
```diff
@@ -0,0 +1,45 @@
+ * and return one of them.
+ */
+@Operator
+public abstract class CoreOps {
```
I renamed it to `Helpers`, but that's also not a great name.

Craigacp commented on tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/CoreOps.java:
```diff
@@ -0,0 +1,45 @@
+        if (exEnv instanceof Graph) {
+            Graph graph = (Graph) exEnv;
+            graph.addInitializer(assignOp);
+        }
```
It doesn't work in Eager mode, I've changed it to throw an `IllegalArgumentException` and documented that it doesn't work.
